### PR TITLE
#10 Refactor to enable unit testing.

### PR DESCRIPTION
**Pull request by michaelw85**
I wanted to add some functionality to the plugin but I first wanted to make the code a bit more readable and testable.

I might have fixed issue #7 I noticed some code using full text for searching instead of the text where comments are stripped. Used the version without comments. Also noticed some mistakes in the backwards search functionality, replaced it with native js.

### CODE REVIEW

**Comment by anacierdem** [comment]
Thank you for your improvements! I will review and merge as soon as I have some free time.

**Comment by michaelw85** [comment]
@anacierdem I just pushed a change, refactoring getting the params and paths. By accident this resolved an issue adding support for CommonJS style require statements (this is the feature I wanted to add to the extension because it was not working)

File: referenceProvider.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    provideDefinition(document, position) {
+        const fullText = document.getText();
+        const currentFilePath = document.fileName;
+        const range = document.getWordRangeAtPosition(position);
+
+        let childWord = "";
+        let moduleList = {};
+
+        if(range) {
+            const textAtCarrot = document.getText(range);
```
Comment (anacierdem) [inline referenceProvider.js]: textAtCaret

File: referenceProvider.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+                return search.forModule(currentFilePath, modulePath, searchFor, stopSearchingFurther);
+            } else { //word is not a module
+                return new Promise(resolve => {
+                    let continueFrom;
+
+                    let dotPosition = document.offsetAt(new vscode.Position(range._start._line, range._start._character-1));
+
+                    //Do backwards search for a dot
+                    dotPosition = fullTextWithoutComments.lastIndexOf('.', dotPosition);
```
Comment (anacierdem) [inline referenceProvider.js]: The old method also returns false if there is a character other than a space or tab or new line. Are you sure this matches the same functionality? I doubt it.

File: referenceProvider.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+                        if(document.getText(document.getWordRangeAtPosition(constructors[0].range._start)) == textAtCarrot) {
+                            resolve(undefined);
+                            return;
+                        } else {
+                            continueFrom = constructors[0].range._start;
+                        }
+                    } else if(haveParent) { //Have a parent - follow it
+                        const propertyParentPosition = document.positionAt(dotPosition);
+                        let bracketPosition = document.offsetAt(propertyParentPosition);
+                        //Do backwards search for a ")"
```
Comment (anacierdem) [inline referenceProvider.js]: The old method also returns false if there is a character other than a space or tab or new line. Are you sure this matches the same functionality? I doubt it.

File: referenceProvider.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
+const vscode = require('vscode');
+const parser = require('./parser');
```
Comment (anacierdem) [inline referenceProvider.js]: On the second thought, I didnt like this new module scheme. The original file is already small. Could you only provide fixes on the main file? I will then think of a module structure.

File: test/parser.test.js [hunk 5]
```diff
@@ -0,0 +1,47 @@
+const assert = require('assert');
+const parser = require('../parser');
+
+suite('Parser', () => {
+
+    test('getModulesWithPathFromRequireOrDefine should return object with module path and name', () => {
```
Comment (anacierdem) [inline test/parser.test.js]: These only cover very limited test cases.

**Comment by michaelw85** [comment]
Hi anacierdem,

I will try to resolve your comments asap.
Can you provide me with the files you use to test? I took a requirejs test project which does not cover all the cases you mentioned.

In response to your remark on separating into modules, the extension file is what I consider big. It's hard to understand what is going on because of it being a big function with lots of if/else/elseif statement and nested functions cluttering the file. It's also really challenging to unit test the original setup hence the split. I know I only covered a couple really basic tests, but its a start to get at least the happy flow tested. I will try to add more when I get a better understanding of what each piece of code should be doing.

Regards,
Michael

**Comment by anacierdem** [comment]
Just dont use modules for the time being (or at least only do a commit ONLY including module migration o/w it is really hard to follow changes). I will try to provide the test files, even include them on the repo.

**Comment by michaelw85** [comment]
Ok, I will revert part of what I did so I can introduce changes slowly to make it easier to review.

**Comment by anacierdem** [comment]
Committed test files, FYI

**Comment by michaelw85** [comment]
Great! I wanted to work on the changes tonight but I have no time atm. Will try to continue upcoming weekend.

**Comment by anacierdem** [comment]
OK, I suggest working over the latest version.

File: referenceProvider.js (hunk 4 above)
Comment (michaelw85) [inline referenceProvider.js]: Will keep everything in 1 file

File: referenceProvider.js (hunk 1 above)
Comment (michaelw85) [inline referenceProvider.js]: Will update.

File: referenceProvider.js (hunk 2 above)
Comment (michaelw85) [inline referenceProvider.js]: I will revert this change.

File: test/parser.test.js (hunk 5 above)
Comment (michaelw85) [inline test/parser.test.js]: Until I get a better understanding of the code this is the bare minimum/best I can do. Lets get some tests in place then we can add more soon :)

**Comment by michaelw85** [comment]
@anacierdem I think I'm done. I tried to cover as much functionality as I could in unit tests. All functionality is still in one file, I just had to reorder and expose the functions to make them testable. I also applied refactoring var to const/let and replaced some function with arrow functions. 
Functionality should be fully intact, manually tested using the files you provided.

File: extension.js [hunk 6]
```diff
@@ -1,328 +1,346 @@
-var vscode = require('vscode');
-var path = require('path');
-
-const JS_MODE = { scheme: 'file' };
```
Comment (anacierdem) [inline extension.js]: Dont we need this?

**Comment by anacierdem** [comment]
Thank you for your through work, I really appreciate that. Merged but it will take a few days before I can get a release. Haven't tested changes myself yet, but they all seem good to me.

**Comment by anacierdem** [comment]
I think there is a problem with onlyNavigateToFile = false. But I will fix this.

**Comment by michaelw85** [comment]
Thanks @anacierdem! I don't think you need to create a new release functionality did not change. Now that it is merged I will try to do some more work on the extension.
